apollo-federation/src/composition/satisfiability/satisfiability_error.rs

```diff
@@ -0,0 +1,466 @@
+                    is_part_of_provides: _,
+                } => {
+                    let parent_type_pos = field_definition_position.parent();
+                    let Some(field) = FieldSelection::from_field(
+                        &build_witness_field(schema, field_definition_position)?,
+                        &parent_type_pos,
+                        &Default::default(),
+                        schema,
+                        &|| Ok(()), // never cancels
+                    )?
+                    else {
+                        bail!("build_witness_next_step: field is None");
+                    };
+                    let field = field.with_updated_selection_set(sub_selection);
+                    (parent_type_pos.clone(), field.into())
+                }
+                _ => {
+                    // Witnesses are build from a path on the supergraph, so we shouldn't have any of those edges.
+                    bail!("Invalid edge {edge} found in supergraph path");
+                }
+            };
+            Ok(Some(SelectionSet::from_selection(parent_type, selection)))
+        }
+    }
+}
+
+fn build_witness_field(
+    schema: &ValidFederationSchema,
+    field_definition_position: &FieldDefinitionPosition,
+) -> Result<executable::Field, FederationError> {
+    // let parent_type_pos = field_definition_position.parent();
+    let field_def = field_definition_position.get(schema.schema())?;
+    let result = executable::Field::new(field_def.name.clone(), field_def.node.clone());
+    let args = field_def
+        .arguments
+        .iter()
+        .filter_map(|arg_def| {
+            if !arg_def.is_required() {
+                return None;
+            }
+            let arg_value = match generate_witness_value(schema, arg_def) {
+                Ok(value) => value,
+                Err(e) => {
+                    return Some(Err(e));
+                }
+            };
+            Some(Ok(Node::new(ast::Argument {
+                name: arg_def.name.clone(),
+                value: arg_value,
+            })))
+        })
+        .collect::<Result<Vec<_>, _>>()?;
+    if args.is_empty() {
+        Ok(result)
+    } else {
+        Ok(result.with_arguments(args))
+    }
+}
+
+fn generate_witness_value(
+    schema: &ValidFederationSchema,
+    value_def: &ast::InputValueDefinition,
+) -> Result<Node<ast::Value>, FederationError> {
+    // Note: We always generate a non-null value, even if the argument is nullable.
+    let value = match value_def.ty.as_ref() {
+        executable::Type::Named(type_name) | executable::Type::NonNullNamed(type_name) => {
+            let type_pos = schema.get_type(type_name.clone())?;
+            match type_pos {
+                TypeDefinitionPosition::Scalar(scalar_type_pos) => {
+                    match scalar_type_pos.type_name.as_str() {
+                        "Int" => ast::Value::Int(0.into()),
+                        "Float" => ast::Value::Float((3.1).into()),
```

Comment:
The JS code used `3.14` here; it's fine to change it (e.g. because `3.1` is smaller and still gets the point across that it's an arbitrary float value), but it's worth leaving a `PORT_NOTE` here to avoid confusion.